### ISSUE TITLE
question: does caffe2 support 3D images?

### ISSUE THREAD

**Issue by emoebel** [facebookarchive/caffe2#587]
Hello,
I have a simple question: does caffe2 support 3D images? I'm not speaking about channel dimension, I'm speaking about 3 spatial dimensions, like in medical tomography for example. This requires 3D conv layers and 3D pooling layers. I know that [elektronn](http://elektronn.org) and [keras](https://keras.io) support 3D images, but last I checked caffe did not. Now that caffe2 is out, I was wondering if this is the case now?
Best,
Lazyleaf

**Comment by pietern** [comment]
@asaadaldien worked on N-D convolutions so can best answer this.

**Comment by asaadaldien** [comment]
Yes we do have an implementation for 3D (convolution, max pooling, average pooling and spatial batch normalization).
For convolution you can use model.ConvND and pass 3d kernel, strides and padding parameters here is an example : 
`model.ConvND("x", "y", kernels=[3,3,3], strides=[1,1,1], pads=[1,1,1]*2)`
 `model.MaxPool("y", "pooled_y", kernels=[2,2,2], strides=[1,1,1], pads=[1,1,1]*2)`

**Comment by williford** [comment]
Actually, caffe 1 has supported ND convolutions since:
https://github.com/BVLC/caffe/pull/2049
